yt_dlp/update.py

```diff
@@ -146,179 +164,292 @@ def _sha256_file(path):
     return h.hexdigest()
 
 
-class Updater:
-    _exact = True
+def _make_label(origin, tag, version=None):
+    if '/' in origin:
+        channel = _INVERSE_UPDATE_SOURCES.get(origin, origin)
+    else:
+        channel = origin
+    label = f'{channel}@{tag}'
+    if version and version != tag:
+        label += f' build {version}'
+    if channel != origin:
+        label += f' from {origin}'
+    return label
 
-    def __init__(self, ydl, target=None):
-        self.ydl = ydl
 
-        self.target_channel, sep, self.target_tag = (target or CHANNEL).rpartition('@')
-        # stable => stable@latest
-        if not sep and ('/' in self.target_tag or self.target_tag in UPDATE_SOURCES):
-            self.target_channel = self.target_tag
-            self.target_tag = None
-        elif not self.target_channel:
-            self.target_channel = CHANNEL.partition('@')[0]
+@dataclass
+class UpdateInfo:
+    """
+    Update target information
+
+    Can be created by `query_update()` or manually.
+
+    Attributes:
+        tag                The release tag that will be updated to. If from query_update,
+                        the value is after API resolution and update spec processing.
+                        The only property that is required.
+        version            The actual numeric version (if available) of the binary to be updated to,
+                        after API resolution and update spec processing. (default: None)
+        requested_version  Numeric version of the binary being requested (if available),
+                        after API resolution only. (default: None)
+        commit             Commit hash (if available) of the binary to be updated to,
+                        after API resolution and update spec processing. (default: None)
+                        This value will only match the RELEASE_GIT_HEAD of prerelease builds.
+        binary_name        Filename of the binary to be updated to. (default: current binary name)
+        checksum           Expected checksum (if available) of the binary to be
+                        updated to. (default: None)
+    """
+    tag: str
+    version: str | None = None
+    requested_version: str | None = None
+    commit: str | None = None
+
+    binary_name: str | None = _get_binary_name()
+    checksum: str | None = None
 
-        if not self.target_tag:
-            self.target_tag = 'latest'
+    _has_update = True
+
+
+class Updater:
```

Comment:
Might be worth writing some unit tests for the new updater

Reply:
Agreed; the plan was to do so in the near future, and the new updater tries to be more unit testable. Not sure if it's necessary in this PR though, since we are actually changing its behavior: the old updater's logic was already incorrect in places, and its behavior incompatible with what we wanted to achieve with `--update-to`

Reply:
I would imagine changing something critical like this should warrant testing. doing it later will be too late

Reply:
Are are intergration testing on `actions-test` as well as bashonlys fork

Reply:
See https://github.com/yt-dlp/yt-dlp/pull/8357/commits/27166cbf3b343ad7f25f9512bf75dfaef4e7714e